### DISCUSSION TITLE
Why is there no on_match callback on rulers

### DISCUSSION THREAD

**Discussion by rafelafrance** [explosion/spaCy#12265] [New Features & Project Ideas]
I just started resurrecting some old spaCy rule-based parsers and I noticed some great new-to-me rulers that do not have an `on_match` callback. The entity and span rulers would really fit well with my particular workflow if they had callbacks.

There's probably a good reason for this aside from the matcher/ruler naming difference. I'm curious what it is.

-----
My use case is pretty common, I build up bigger patterns from smaller patterns and capture details along the way. For instance:
1. Given a string like: `Leaf (12-)23-34 × 45-56 cm`
2. I capture the 2 range patterns `(12-)23-34` & `45-56` using a pattern that is roughly:
    - `{"TEXT": {"IN": " ( [ ".split()}}`
    - `{"TEXT": {"REGEX": r"^{(\d+\.?\d*)}$"}}`
    - `{"TEXT": {"IN": " – - –– -- ".split()}}`
    - `{"TEXT": {"IN": " ) ] ".split()}}`
    - etc.
3. In an `on_match` callback put the values  `min=12`, `low=23`, `high=34` etc. in the range entity's `ent._.data` dict.
4. I then match on a "higher" pattern that boils down to: `range cm? x range cm`
    - I find this notation easier to debug.
5. In another callback I pull the lower values into a size entity's `ent._.data` and match them to the dimensions etc.

Steps 3 & 5 are the problem. This is simplified of course. And, to be honest, I'm probably not using features that spaCy already provides to their best use.

**Comment by rmitsch** [comment]
Hey @rafelafrance, the reason for this is that `EntityRuler` and `SpanRuler` have a more narrow, pre-implemented "callback" by default: storing matches in `.ents` or `.spans`. 
However, when using `SpanRuler`, you can set an arbitrary function as `spans_filter` (see [docs](https://spacy.io/api/spanruler#config)). You are free to do pretty much whatever in there, so I could imagine that this might be a viable approach for you.

**Reply by rafelafrance** [reply]
Thank you.

What I have now works, it's not elegant but it works, so I was more interested in why. These tidbits of knowledge help me understand spacy. SpaCy is so useful, you all have done an excellent job.

I'll definitely try using the `spans_filter` to see if that cleans things up a bit and if it's easy to fit into the pile of code I've already written.

**Reply by rmitsch** [reply]
Thanks for the kind words :slightly_smiling_face:  We're happy spaCy is useful to you.
Feel free to open another discussion if you run into problems!
